gap/Forms.gi

```diff
@@ -1,56 +1,77 @@
-            fi;
-            gapForm := HermitianFormByMatrix(formMatrix, field);
+        formMatrix := UnitaryForm(group);
+        if formMatrix = fail then
+            ErrorNoReturn("No preserved unitary form found for <group>");
         fi;
+        gapForm := HermitianFormByMatrix(formMatrix, field);
         newForm := HermitianFormByMatrix(gramMatrix, field);
     fi;
-    # the following if condition can only ever be fulfilled if <group> is an
-    # orthogonal group; there the case of even dimension is problematic since,
-    # in that case, there are two similarity classes of bilinear forms
-    if not WittIndex(gapForm) = WittIndex(newForm) then
-       ErrorNoReturn("The form preserved by <group> must be similar to the form ", 
-                     "described by the Gram matrix <gramMatrix>.");
-    fi;
-    gapToCanonical := BaseChangeHomomorphism(BaseChangeToCanonical(gapForm), 
-                                             field);
-    canonicalToNew := BaseChangeHomomorphism(BaseChangeToCanonical(newForm) ^ (-1), 
-                                             field);
-    result := MatrixGroup(field, canonicalToNew(gapToCanonical(GeneratorsOfGroup(group))));
-    if HasSize(group) then
-        SetSize(result, Size(group));
+    # The Form package has a bug for d = 1 so we need to make this exception
+    if not d = 1 then
+        # the following if condition can only ever be fulfilled if <group> is an
+        # orthogonal group; there the case of even dimension is problematic since,
+        # in that case, there are two similarity classes of bilinear forms
+        if not WittIndex(gapForm) = WittIndex(newForm) then
+            ErrorNoReturn("The form preserved by <group> must be similar to the form ", 
+                          "described by the Gram matrix <gramMatrix>.");
+        fi;
+        gapToCanonical := BaseChangeHomomorphism(BaseChangeToCanonical(gapForm), 
+                                                 field);
+        canonicalToNew := BaseChangeHomomorphism(BaseChangeToCanonical(newForm) ^ (-1), 
+                                                 field);
+        result := MatrixGroup(field, canonicalToNew(gapToCanonical(GeneratorsOfGroup(group))));
+    
+        # Set useful attributes
+        if HasSize(group) then
+            SetSize(result, Size(group));
+        fi;
+    else
+        # replaces the Witt index check above
+        if (gramMatrix[1, 1] = 0 and formMatrix[1, 1] <> 0) 
+           or (gramMatrix[1, 1] <> 0 and formMatrix[1, 1] = 0) then
+            ErrorNoReturn("The form preserved by <group> must be similar to the",
+                          " form described by the Gram matrix <gramMatrix>.");
+        fi;
+        result := group;
+    fi;
+
+    if type = "S" or type = "O" then
+        SetInvariantBilinearForm(result, rec(matrix := gramMatrix));
+    else
+        SetInvariantSesquilinearForm(result, rec(matrix := gramMatrix));
     fi;
```

Comment:
So I guess this means even characteristic is not (yet) fully supported here, right? Because no quadratic form is set

Reply:
Yes, this function only does bilinear forms, not quadratic forms, i.e. the orthogonal case only in odd characteristic